grafanalib/elasticsearch.py

```diff
@@ -0,0 +1,178 @@
+
+    def to_json_data(self):
+        return {
+                'filters': self.filters
+               }
+
+
+@attr.s
+class FiltersAgg(object):
+    field = attr.ib(default="time_iso8601", validator=instance_of(str))
+    id = attr.ib(default=0, validator=instance_of(int))
+    settings = attr.ib(default=FiltersAggSettings())
+    type = attr.ib(default="filters", validator=instance_of(str))
+
+    def to_json_data(self):
+        return {
+                'field': self.field,
+                'id': str(self.id),
+                'settings': self.settings,
+                'type': self.type
+                }
+
+
+@attr.s
+class TermsAggSettings(object):
+    min_doc_count = attr.ib(default=1, validator=instance_of(int))
+    order = attr.ib(default="desc", validator=instance_of(str))
+    order_by = attr.ib(default="_term", validator=instance_of(str))
+    size = attr.ib(default=0, validator=instance_of(int))
+
+    def to_json_data(self):
+        return {
+                'min_doc_count': self.min_doc_count,
+                'order': self.order,
+                'order_by': self.order_by,
+                'size': self.size
+                }
+
+
+@attr.s
+class TermsAgg(object):
+    field = attr.ib(validator=instance_of(str))
+    id = attr.ib(default=0, validator=instance_of(int))
+    settings = attr.ib(default=TermsAggSettings())
+    type = attr.ib(default="terms", validator=instance_of(str))
+
+    def to_json_data(self):
+        return {
+                'field': self.field,
+                'id': str(self.id),
+                'settings': self.settings,
+                'type': self.type
+                }
+
+
+@attr.s
+class ElasticsearchTarget(object):
+    """Generates Elasticsearch target JSON structure.
+
+    Grafana docs on using Elasticsearch:
+    http://docs.grafana.org/features/datasources/elasticsearch/
+    Elasticsearch docs on querying or reading data:
+    https://www.elastic.co/guide/en/elasticsearch/reference/current/index.html
+
+    :param alias: legend alias
+    :param bucketAggs: Group by query aggregators
+    :param metrics: Elasticsearch metric
+    :param query: query
+    :param refId: target reference id
+    """
+
+    alias = attr.ib(default=None)
+    bucket_aggs = attr.ib(default=[DateTimeAgg()])
+    metrics = attr.ib(default=[CountMetric()])
+    query = attr.ib(default="", validator=instance_of(str))
+    refId = attr.ib(default="", validator=instance_of(str))
+
+    def auto_bucket_aggs_id(self):
+        ids = set([agg.id for agg in self.bucket_aggs if agg.id])
+        auto_ids = (i for i in itertools.count(1) if i not in ids)
+
+        def set_id(agg):
+            if agg.id:
+                return agg
+
+            agg.id = next(auto_ids)
+            return agg
+
+        return list(map(set_id, self.bucket_aggs))
+
+    def to_json_data(self):
+        return {
+            'alias': self.alias,
+            'bucketAggs': self.auto_bucket_aggs_id(),
```

Comment:
I agree it should be consistent with `Dashboard`.

My thought was that it's not obvious that `auto_panel_ids()` needs to be called. If you don't call it you get a broken dashboard without any hint what's wrong.
There is no error message, no output neither from grafana nor from grafanalib that it's because of problems with id enumeration.

So I think it would be better to do the enumeration transparent for the user, that it simply works.